### PR TITLE
9 async lesiaturuta

### PR DESCRIPTION
**Pull request by lesiaturuta**

### CODE REVIEW

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 1]
```diff
@@ -1,273 +1,21 @@
...
-        ThumbnailAltText: 'Charizard',
-        ThumbnailImage: 'https://assets.pokemon.com/assets/cms2/img/pokedex/detail/006.png',
-        id: 6,
-        type: [
-            'fire',
-            'dragon',
-        ],
-    },
-];
+let pokemons = [];
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:2]: Hey, @lesiaturuta, thanks for a good homework. 
Here is bug for you:
1. Type something in the search bar and hit enter
2. Choose a radio btn
Expected result: only pokemons matched the search text will be rendered
Actual result: all cards are shown.

image

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 2]
```diff
@@ -1,273 +1,21 @@
...
 
 const cardsEl = document.querySelector('[data-cards]');
 const selectSearchEl = document.querySelector('[data-filter-select-search]');
-const selectSearchRadioOnlyNewEl = document.querySelector('[data-filter-new]');
-const selectSearchRadioFavoritesEl = document.querySelector('[data-filter-favorites]');
+const selectSearchRadioHighEl = document.querySelector('[data-filter-high]');
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:6]: Why do you add `selectSerch` to the names of all these elements? 
`selectSearchRadioHighEl` is not a search
`selectSearchEl` is not a select
...

You can pick proper name, like e.g. `searchEl`or `searchFilterEl`, `highFilterEl`

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 3]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
+        filterPokemonsArray = [...pokemons];
+        typesPokemons = updateTypes(pokemons);
+        renderSelect();
+        loadingEl.classList.add('display-none');
+    } catch (e) {
+        alert('Error url, please update this page');
+        console.error(e);
+    }
+
+    // disabled for html elements
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:126]: Cloud you remove these comments please

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 4]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
 
-const init = () => {
-// disabled for html elements
-    selectSearchEl.disabled = true;
-    selectSearchRadioOnlyNewEl.disabled = true;
-    selectSearchRadioFavoritesEl.disabled = true;
+const filterPokemonsByType = (type) => pokemons.filter((pokemon) => pokemon.type.includes(type));
+
+const init = async () => {
+    // get pokemons
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:112]: Don't you think, `init` function is too big?
You event add a comment to describe what is going on in this piece of code. Proly you can move it to the function with this very name `getPokemons`

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 5]
```diff
@@ -315,6 +66,34 @@ const cleanElement = (el) => {
...
+        optionEl.text = type;
+        selectSearchEl.add(optionEl);
+    });
+};
+
+const sortPokemonsUp = (array) => array.sort((a, b) => a.height - b.height);
+
+const sortPokemonsDown = (array) => array.sort((a, b) => b.height - a.height);
+
+const updateTypes = (array) => {
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:89]: If a function returns something, better to show it in its name: `getPokemonTypes`.
 And an `array` is too common name for this parameter. If you are working with pokemons, name it `pokemons`

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 6]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
+    // selectSearchEl.disabled = true;
+    // selectSearchRadioHighEl.disabled = true;
+    // selectSearchRadioLowEl.disabled = true;
     // selectSearchInputSearchEl.disabled = true;
     // selectSearchInputButtonEl.disabled = true;
 
     selectSearchCheckboxBigEl.addEventListener('change', (e) => {
         checkboxBig = e.target.checked;
         filterPokemonsArray = filterPokemons(checkboxBig, checkboxSmall);
+        typesPokemons = updateTypes(pokemons);
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:136]: Couldn't you make one handler for both chackboxes?
Add a proper name, lets say `onSizeFilterChange` and get `checked` state of both checkboxes in it.

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 7]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
+
+const init = async () => {
+    // get pokemons
+    try {
+        const url = 'https://my-json-server.typicode.com/electrovladyslav/pokemon-json-server/pokemons';
+        const response = await fetch(url);
+        pokemons = await response.json();
+        filterPokemonsArray = [...pokemons];
+        typesPokemons = updateTypes(pokemons);
+        renderSelect();
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:119]:
```suggestion
        populateSelect();
```

**Review by vladyslav-yermolin-moc — CHANGES_REQUESTED** [review]

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 8]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
         checkboxSmall = e.target.checked;
         filterPokemonsArray = filterPokemons(checkboxBig, checkboxSmall);
+        typesPokemons = updateTypes(pokemons);
+        renderSelect();
         cleanElement(cardsEl);
         renderCards(cardsEl, filterPokemonsArray);
     });
 
+    selectSearchEl.addEventListener('change', (e) => {
+        filterPokemonsArray = filterPokemonsByType(e.target.value);
```
[the remark this answers is not in the document]
Comment (lesiaturuta) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js]: fixed a lot of code

**Comment by lesiaturuta** [comment]
fixed a lot of code

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 9]
```diff
@@ -309,58 +66,138 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
 
-const init = () => {
-// disabled for html elements
-    selectSearchEl.disabled = true;
-    selectSearchRadioOnlyNewEl.disabled = true;
-    selectSearchRadioFavoritesEl.disabled = true;
-    // selectSearchInputSearchEl.disabled = true;
-    // selectSearchInputButtonEl.disabled = true;
+const init = async () => {
+    try {
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:160]: If you have `try catch ` block in the `getPokemons` function, you can skip it here

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 5 above)
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:98]: still need to fix

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 10]
```diff
@@ -309,58 +66,138 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
 
-const filterPokemons = (checkBig, checkSmall) => {
-    let newPokemons = pokemons;
-    if (checkBig && checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight < 100 && pokemon.weight > 50);
-    } else if (checkBig && !checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight > 100);
-    } else if (!checkBig && checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight < 50);
+const getFilterPokemons = () => {
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:125]:
```suggestion
const getFilteredPokemons = () => {
```

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 6 above)
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js]: ok, up to you))

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/style.css [hunk 11]
```diff
@@ -570,5 +570,5 @@ body {
 }
 
 .display-none {
-  display: none;
+  display: none !important;
```
[the remark this answers is not in the document]
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/style.css:573]: Still need to fix
You can add `z-index: -1` for exapmle

[the remark this answers is not in the document]
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/style.css:573]: Or you can remove spinner element, not just add class
Or you can add `hidden: true`

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 12]
```diff
@@ -309,58 +66,138 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
+    } catch (e) {
+        alert('Error url, please update this page');
+        console.error(e);
+    }
+    typesPokemons = updateTypes(pokemons);
+    renderSelect(typeValue);
+    loadingEl.classList.add('display-none');
 
-    selectSearchCheckboxBigEl.addEventListener('change', (e) => {
+    checkboxBigEl.addEventListener('change', (e) => {
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:170]: I still think, that `init` function is too long.
Could you make a `addHandlers` function and move to it all logic related to `addEventListener`

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 13]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
+    // selectSearchRadioHighEl.disabled = true;
+    // selectSearchRadioLowEl.disabled = true;
     // selectSearchInputSearchEl.disabled = true;
     // selectSearchInputButtonEl.disabled = true;
 
     selectSearchCheckboxBigEl.addEventListener('change', (e) => {
         checkboxBig = e.target.checked;
         filterPokemonsArray = filterPokemons(checkboxBig, checkboxSmall);
+        typesPokemons = updateTypes(pokemons);
+        renderSelect();
```
[the remark this answers is not in the document]
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js]: Once again, do you think, that if a user selects one value, you should update the options of select? I my opinion, the user should see all options in this case
image

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 14]
```diff
@@ -309,58 +66,138 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
+    array.forEach((pokemon) => pokemon.type.forEach((type) => {
+        if (!types.includes(type)) types.push(type);
+    }));
+    return types;
+};
+
+const getPokemonsByType = (type, arrayPokemons) => arrayPokemons
+    .filter((pokemon) => pokemon.type
+        .includes(type));
+const getPokemonsByValet = (text, arrayPokemons) => arrayPokemons
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:109]:
```suggestion
const getPokemonsBySearchName = (text, arrayPokemons) => arrayPokemons
```
`Valet` - what does it mean?

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/style.css (hunk 11 above)
[the remark this answers is not in the document]
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/style.css:573]: Or, probably the best solution, you can add a cascade
```css
.loading.display-none
```

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 14 above)
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:109]: you can add a logic of empty search string here
```
if (text === '') return arrayPokemon
```

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 15]
```diff
@@ -309,58 +66,138 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
-    let newPokemons = pokemons;
-    if (checkBig && checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight < 100 && pokemon.weight > 50);
-    } else if (checkBig && !checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight > 100);
-    } else if (!checkBig && checkSmall) {
-        newPokemons = pokemons.filter((pokemon) => pokemon.weight < 50);
+const getFilterPokemons = () => {
+    let newPokemons = [];
+    if (searchValue && typeValue) {
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:127]: it is so easy to lost in all these if-else
could you make some auxiliary functions, to make this block simpler?
```
newPokemons = [... pokemons]

newPokemons = getPokemonsByName(newPokemon)  // logic of empty search is inside

newPokemons = getPokemonsByType(newPokemons)  // logic of not selected any type inside

newPokemons = getPokemonsByWeight(newPokemons) // inside you get checked size checkboxes and filter pokemons by them

newPokemons = sortByHeight(newPokemons) 
```

**Review by vladyslav-yermolin-moc — CHANGES_REQUESTED** [review]

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 9 above)
Comment (lesiaturuta) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:160]: fixed

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 10 above)
Comment (lesiaturuta) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:125]: fixed

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js (hunk 12 above)
Comment (lesiaturuta) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:170]: fixed

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 16]
```diff
@@ -327,28 +106,66 @@ const filterPokemons = (checkBig, checkSmall) => {
...
+    });
+
+    selectSearchRadioLowEl.addEventListener('change', () => {
+        cleanElement(cardsEl);
+        renderCards(cardsEl, sortPokemonsDown(filterPokemonsArray));
+    });
+
     selectSearchInputButtonEl.addEventListener('click', (e) => {
         e.preventDefault();
         // eslint-disable-next-line max-len
```
[the remark this answers is not in the document]
Comment (vladyslav-yermolin-moc) [reply homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js]: ok, up to U))

File: homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js [hunk 17]
```diff
@@ -309,58 +66,131 @@ const renderCards = (cardContainer, pokemonsArray) => {
...
+    ? arrayPokemons
+        .filter((pokemon) => pokemon.name
+            .toLocaleLowerCase()
+            .includes(searchValue.toLocaleLowerCase()))
+    : arrayPokemons);
+
+const getPokemonsByWeight = (arrayPokemons) => {
+    let newPokemons = [...arrayPokemons];
+    if (checkboxBig && checkboxSmall && newPokemons.length) {
+        newPokemons = newPokemons.filter((pokemon) => pokemon.weight < 100 && pokemon.weight > 50);
```
Comment (vladyslav-yermolin-moc) [inline homeworks/lesia.turua_lesiaturuta/PokemonProject/scripts/pokemon.js:114]:
```suggestion
        newPokemons = newPokemons.filter((pokemon) => pokemon.weight > 100 && pokemon.weight < 50);
```
Please fix in the final version

**Review by vladyslav-yermolin-moc — APPROVED** [review]

**Comment by vladyslav-yermolin-moc** [comment]
Well done, thanks for your fixes👏

**Pull request merged**